### PR TITLE
Warn about side-effect free statements.

### LINKED ISSUE #1816: Warn about some expression statements without side effects
This is a story to implement as much of
https://github.com/ethereum/solidity/issues/1380
as possible already now. It is better to have many false negatives than no warning at all. We can probably re-use the "is pure" analysis for constant variables here.

### LINKED ISSUE #2121: Compiler does not throw error on improper use of `revert` keyword
The `revert` keyword can be used in two ways that may cause confusion. As a function call `revert()` in solidity, or as an opcode `revert` in inline assembly.

The compiler however does not throw an error if used as a keyword in solidity instead of a call.  And the compiled bytecode does not contain the REVERT opcode as might be expected.

    contract Revert{
    
    // Improper successful compilation on invalid use of `revert` keyword
    //  JUMPDEST 			function r1() {\n        rever...
    //  JUMP [out]			function r1() {\n        rever...
    function r1() {
        revert;
    }

    // proper use and function of keyword as function call
    //  JUMPDEST 			function r2() {\n        rever...
    //  PUSH 0			revert()
    //  PUSH 0			revert()
    //  REVERT 			revert()
    function r2() {
        revert();
    }
    
    // proper use and function of keyword in assembly
    //  JUMPDEST 			function r3() {\n        assem...
    //  REVERT 			revert
    function r3() {
        assembly {
            revert
        }
    }
    }

### PR DESCRIPTION
**Pull request by chriseth**
Fixes #2121 
Fixes #1816

### CODE REVIEW

File: libsolidity/analysis/TypeChecker.cpp [hunk 1]
```diff
@@ -1674,8 +1674,8 @@ bool TypeChecker::visit(Identifier const& _identifier)
 	if (auto variableDeclaration = dynamic_cast<VariableDeclaration const*>(annotation.referencedDeclaration))
 		annotation.isPure = annotation.isConstant = variableDeclaration->isConstant();
 	else if (dynamic_cast<MagicVariableDeclaration const*>(annotation.referencedDeclaration))
-		if (auto functionType = dynamic_cast<FunctionType const*>(annotation.type.get()))
-			annotation.isPure = functionType->isPure();
+		if (dynamic_cast<FunctionType const*>(annotation.type.get()))
+			annotation.isPure = true;
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp:1678]: Please check that this change is correct. This is the difference between `f` and `f()`. The first should be pure even if `f` is not a pure function.

File: libsolidity/analysis/StaticAnalyzer.cpp [hunk 2]
```diff
@@ -57,6 +57,13 @@ void StaticAnalyzer::endVisit(FunctionDefinition const&)
 	m_nonPayablePublic = false;
 }
 
+bool StaticAnalyzer::visit(ExpressionStatement const& _statement)
+{
+	if (_statement.expression().annotation().isPure)
+		warning(_statement.location(), "Statement has no effects.");
```
Comment (axic) [inline libsolidity/analysis/StaticAnalyzer.cpp]: Probably singular `no effect` is better.

**Comment by axic** [comment]
Can you add a test for `var a = revert; a;` ?

**Comment by chriseth** [comment]
@axic such a test would be identical to `uint a; a;`, but I'll add it.

**Comment by chriseth** [comment]
@axic it turns out such code will not trigger the warning, because `a` is not considered pure by the current code - it could be changed in a loop or so. I guess we will still catch most of the cases.

**Comment by chriseth** [comment]
Note that this is not detecting all cases of side-effect-free statements, the rest will be done in https://github.com/ethereum/solidity/issues/1380

File: libsolidity/analysis/TypeChecker.cpp (hunk 1 above)
Comment (pirapira) [reply libsolidity/analysis/TypeChecker.cpp:1678]: Why is this just for functions?  What about other magic variables?

Comment (pirapira) [reply libsolidity/analysis/TypeChecker.cpp:1678]: For instance, `msg` might be pure although `msg.sender` is not pure.

File: test/libsolidity/SolidityNameAndTypeResolution.cpp [hunk 3]
```diff
@@ -3653,54 +3653,56 @@ BOOST_AUTO_TEST_CASE(conditional_with_all_types)
 				// integers
 				uint x;
 				uint y;
-				true ? x : y;
+				uint g = true ? x : y;
```
Comment (pirapira) [inline test/libsolidity/SolidityNameAndTypeResolution.cpp:3656]: Do we have an issue to warn about unread variables?

File: test/libsolidity/SolidityNameAndTypeResolution.cpp [hunk 4]
```diff
@@ -5477,6 +5479,32 @@ BOOST_AUTO_TEST_CASE(using_interface_complex)
...
+	CHECK_WARNING(text, "Statement has no effect.");
+}
+
+BOOST_AUTO_TEST_CASE(pure_statement_in_for_loop)
+{
+	char const* text = R"(
+		contract C {
+			function f() {
+				for (uint x = 0; x < 10; true)
+					x++;
```
Comment (pirapira) [inline test/libsolidity/SolidityNameAndTypeResolution.cpp:5509]: Is this a pure statement?

Comment (chriseth) [reply test/libsolidity/SolidityNameAndTypeResolution.cpp:5509]: No it is not.

File: test/libsolidity/SolidityNameAndTypeResolution.cpp (hunk 3 above)
Comment (chriseth) [reply test/libsolidity/SolidityNameAndTypeResolution.cpp:3656]: Yes, someone else is working on it already.

File: test/libsolidity/SolidityNameAndTypeResolution.cpp (hunk 4 above)
Comment (chriseth) [reply test/libsolidity/SolidityNameAndTypeResolution.cpp:5509]: but the `true` is.

Comment (pirapira) [reply test/libsolidity/SolidityNameAndTypeResolution.cpp:5509]: Ah, OK.

Will you add a test about `for (uint x = 0; true; x++)`?

File: libsolidity/analysis/TypeChecker.cpp (hunk 1 above)
Comment (chriseth) [reply libsolidity/analysis/TypeChecker.cpp:1678]: Sure, but do we really need that? When we do the follow-up story, we can add a positive and a negative list and error whenever both are set to "false".

File: test/libsolidity/SolidityNameAndTypeResolution.cpp (hunk 4 above)
Comment (chriseth) [reply test/libsolidity/SolidityNameAndTypeResolution.cpp:5509]: Added the test.

File: libsolidity/analysis/TypeChecker.cpp (hunk 1 above)
Comment (pirapira) [reply libsolidity/analysis/TypeChecker.cpp:1678]: I don't think positives or negatives are relevant.  I'm seeing only booleans.  If an expression reads, and if an expression writes.

We can also do this in the follow-up.  In that case, I'll add this point in the follow-up issue.

**Review by pirapira — APPROVED** [review]
Looks good to me.  I'm restarting the emscripten test.

**Comment by axic** [comment]
I know it is useless because they are regular functions, but would it make sense having a test for all the critical ones (revert, assert, require, selfdestruct)? Just in case some other future feature would mess them up.

**Comment by chriseth** [comment]
Added the tests.

**Comment by chriseth** [comment]
Travis was a sporadic single failure, windows error was race condition with bytecode upload.

**Comment by pirapira** [comment]
@chriseth can there be an issue about the race condition?

**Pull request merged**